File: src/typescript/src/testing.ts
/**
 * Testing utilities for ModelMesh.
 *
 * Provides a mock client that behaves like the real MeshClient but
 * returns pre-configured responses instead of calling live APIs.
 *
 * @example
 * ```ts
 * import { mockClient, MockResponse } from '@modelmesh/core/testing';
 *
 * const client = mockClient({
 *   responses: [
 *     { content: 'Hello!', model: 'gpt-4o', tokens: 10 },
 *     { content: 'World!', model: 'claude-3', tokens: 15 },
 *   ],
 * });
 *
 * const resp = await client.chat.completions.create({
 *   model: 'test-pool',
 *   messages: [{ role: 'user', content: 'Hi' }],
 * });
 * assert(resp.choices[0].message.content === 'Hello!');
 * assert(client.calls.length === 1);
 * ```
 */

import type { CompletionResponse } from './interfaces/provider';

/** Pre-configured response for the mock client. */
export interface MockResponse {
  /** The text content of the assistant's reply. */
  content?: string;
  /** Model identifier to include in the response. */
  model?: string;
  /** Total token count to simulate. */
  tokens?: number;
  /** Prompt token count (defaults to tokens / 3). */
  promptTokens?: number;
  /** Completion token count (auto-calculated). */
  completionTokens?: number;
  /** Stop reason (default: "stop"). */
  finishReason?: string;
  /**
   * When set, this error is thrown instead of returning a response.
   * Useful for simulating provider failures.
   */
  error?: Error;
  /**
   * Seconds to wait before returning or throwing, simulating
   * network latency. Uses `setTimeout` internally.
   */
  delay?: number;
  /**
   * HTTP status code to simulate (informational only).
   * @default 200
   */
  statusCode?: number;
}

/** Record of a call made to the mock client. */
export interface MockCall {
  /** The model / pool name requested. */
  model: string;
  /** The messages sent. */
  messages: Record<string, unknown>[];
  /** Additional parameters. */
  kwargs: Record<string, unknown>;
  /** The response that was returned. */
  response: CompletionResponse;
}

function toCompletionResponse(mock: MockResponse): CompletionResponse {
  const tokens = mock.tokens ?? 10;
  const promptTokens = mock.promptTokens ?? Math.floor(tokens / 3);
  const completionTokens = mock.completionTokens ?? (tokens - promptTokens);
  return {
    id: `mock-${Math.random().toString(36).slice(2, 10)}`,
    model: mock.model ?? 'mock-model',
    choices: [
      {
        index: 0,
        message: {
          role: 'assistant',
          content: mock.content ?? 'Mock response',
        },
        finishReason: mock.finishReason ?? 'stop',
      },
    ],
    usage: {
      promptTokens,
      completionTokens,
      totalTokens: promptTokens + completionTokens,
    },
    created: Math.floor(Date.now() / 1000),
    object: 'chat.completion',
  };
}

class MockChatCompletions {
  private _responses: MockResponse[];
  private _calls: MockCall[];
  private _index = 0;
  private _failureRate: number;

  constructor(responses: MockResponse[], calls: MockCall[], failureRate = 0) {
    this._responses = responses;
    this._calls = calls;
    this._failureRate = failureRate;
  }

  async create(params: {
    model: string;
    messages?: Record<string, unknown>[];
    stream?: boolean;
    [key: string]: unknown;
  }): Promise<CompletionResponse> {
    const messages = params.messages ?? [];
    const mock =
      this._index < this._responses.length
        ? this._responses[this._index++]
        : this._responses[this._responses.length - 1] ?? {};

    // Apply per-response delay
    if (mock.delay && mock.delay > 0) {
      await new Promise((resolve) => setTimeout(resolve, mock.delay! * 1000));
    }

    // Apply global random failure rate (chaos testing)
    if (this._failureRate > 0 && Math.random() < this._failureRate) {
      throw new Error('Simulated random failure (chaos testing)');
    }

    // Apply per-response error injection
    if (mock.error) {
      throw mock.error;
    }

    const response = toCompletionResponse(mock);
    const { model, messages: _msgs, stream: _s, ...kwargs } = params;
    this._calls.push({ model, messages, kwargs, response });
    return response;
  }
}

class MockChatNamespace {
  readonly completions: MockChatCompletions;

  constructor(responses: MockResponse[], calls: MockCall[], failureRate = 0) {
    this.completions = new MockChatCompletions(responses, calls, failureRate);
  }
}

class MockModelsNamespace {
  list() {
    return { data: [], object: 'list' };
  }
}

/**
 * A mock MeshClient for testing.
 *
 * Behaves like the real MeshClient but returns pre-configured
 * responses. Records all calls for assertion.
 */
export class MockClient {
  /** List of recorded calls. */
  readonly calls: MockCall[] = [];
  readonly chat: MockChatNamespace;
  readonly models: MockModelsNamespace;

  constructor(responses?: MockResponse[], failureRate = 0) {
    const resps = responses ?? [{}];
    this.chat = new MockChatNamespace(resps, this.calls, failureRate);
    this.models = new MockModelsNamespace();
  }

  /** Close the mock client (no-op, for API parity with real MeshClient). */
  close(): void {
    // No-op — nothing to clean up in the mock client.
  }

  poolStatus(pool?: string | null): Record<string, unknown> {
    return {
      'mock-pool': {
        active: 1,
        standby: 0,
        total: 1,
        currentModel: 'mock-model',
      },
    };
  }

  activeProviders(): string[] {
    return ['mock-provider'];
  }

  describe(pool?: string | null): string {
    return 'Pool "mock-pool" (strategy: mock)\n  -> mock-model [mock-provider] (active)';
  }

  explain(params?: { model?: string; [key: string]: unknown }): Record<string, unknown> {
    return {
      poolName: 'mock-pool',
      strategy: 'mock',
      capability: 'mock',
      selectedModel: 'mock-model',
      candidates: [
        { modelId: 'mock-model', providerId: 'mock-provider', status: 'active' },
      ],
      reason: 'Mock selection',
    };
  }
}

/**
 * Create a mock MeshClient for testing.
 *
 * @param options - Configuration for the mock client.
 * @param options.responses - Pre-configured responses to cycle through.
 *   Individual responses can specify `error`, `delay`, or `statusCode`
 *   for per-request behavior simulation.
 * @param options.failureRate - Probability (0.0 to 1.0) that any request
 *   randomly throws an Error for chaos testing. Applied before
 *   per-response error checks.
 * @returns A MockClient that can be used in place of a real MeshClient.
 *
 * @example
 * ```ts
 * // Simulate errors
 * const client = mockClient({
 *   responses: [
 *     { content: 'OK' },
 *     { error: new Error('Rate limited') },
 *     { content: 'Recovered' },
 *   ],
 * });
 *
 * // Chaos testing (20% random failures)
 * const client2 = mockClient({
 *   responses: [{ content: 'Response' }],
 *   failureRate: 0.2,
 * });
 * ```
 */
export function mockClient(options?: {
  responses?: MockResponse[];
  failureRate?: number;
}): MockClient {
  return new MockClient(options?.responses, options?.failureRate);
}

File: src/python/modelmesh/testing.py
"""Testing utilities for ModelMesh.

Provides a mock client that behaves like the real ``MeshClient`` but
returns pre-configured responses instead of calling live APIs.
Designed for unit testing applications that use ModelMesh.

Usage::

    from modelmesh.testing import mock_client, MockResponse

    client = mock_client(responses=[
        MockResponse(content="Hello!", model="gpt-4o", tokens=10),
        MockResponse(content="World!", model="claude-3", tokens=15),
    ])

    resp = client.chat.completions.create(
        messages=[{"role": "user", "content": "Hi"}],
        model="test-pool",
    )
    assert resp.choices[0].message.content == "Hello!"
    assert len(client.calls) == 1
"""
from __future__ import annotations

import time
import uuid
from dataclasses import dataclass, field
from typing import Optional

from modelmesh.interfaces.provider import (
    ChatMessage,
    CompletionChoice,
    CompletionResponse,
    TokenUsage,
)


@dataclass
class MockResponse:
    """Pre-configured response for the mock client.

    Attributes:
        content: The text content of the assistant's reply.
        model: Model identifier to include in the response.
        tokens: Total token count to simulate.
        prompt_tokens: Prompt token count (defaults to tokens // 3).
        completion_tokens: Completion token count (auto-calculated).
        finish_reason: Stop reason (default: "stop").
    """

    content: str = "Mock response"
    model: str = "mock-model"
    tokens: int = 10
    prompt_tokens: Optional[int] = None
    completion_tokens: Optional[int] = None
    finish_reason: str = "stop"

    def to_completion_response(self) -> CompletionResponse:
        """Convert to a ``CompletionResponse``."""
        prompt = self.prompt_tokens if self.prompt_tokens is not None else self.tokens // 3
        completion = (
            self.completion_tokens
            if self.completion_tokens is not None
            else self.tokens - prompt
        )
        return CompletionResponse(
            id=f"mock-{uuid.uuid4().hex[:8]}",
            model=self.model,
            choices=[
                CompletionChoice(
                    index=0,
                    message=ChatMessage(
                        role="assistant",
                        content=self.content,
                    ),
                    finish_reason=self.finish_reason,
                )
            ],
            usage=TokenUsage(
                prompt_tokens=prompt,
                completion_tokens=completion,
                total_tokens=prompt + completion,
            ),
            created=int(time.time()),
            object="chat.completion",
        )


@dataclass
class MockCall:
    """Record of a call made to the mock client.

    Attributes:
        model: The model / pool name requested.
        messages: The messages sent.
        kwargs: Additional parameters.
        response: The response that was returned.
    """

    model: str = ""
    messages: list = field(default_factory=list)
    kwargs: dict = field(default_factory=dict)
    response: Optional[CompletionResponse] = None


class _MockChatCompletions:
    """Mock chat.completions namespace."""

    def __init__(self, responses: list[MockResponse], calls: list[MockCall]) -> None:
        self._responses = responses
        self._calls = calls
        self._index = 0

    def create(
        self,
        *,
        model: str = "test",
        messages: list[dict] | None = None,
        stream: bool = False,
        **kwargs,
    ) -> CompletionResponse:
        """Return the next pre-configured response."""
        messages = messages or []

        if self._index < len(self._responses):
            response = self._responses[self._index].to_completion_response()
            self._index += 1
        else:
            # Cycle back to last response if exhausted
            response = self._responses[-1].to_completion_response() if self._responses else MockResponse().to_completion_response()

        call = MockCall(
            model=model,
            messages=messages,
            kwargs=kwargs,
            response=response,
        )
        self._calls.append(call)
        return response


class _MockChatNamespace:
    """Mock chat namespace."""

    def __init__(self, responses: list[MockResponse], calls: list[MockCall]) -> None:
        self.completions = _MockChatCompletions(responses, calls)


class _MockModelsNamespace:
    """Mock models namespace."""

    def list(self):
        return type("ModelList", (), {"data": [], "object": "list"})()


class MockClient:
    """A mock MeshClient for testing.

    Behaves like the real ``MeshClient`` but returns pre-configured
    responses. Records all calls for assertion.

    Attributes:
        calls: List of :class:`MockCall` records.
        chat: Chat namespace with ``completions.create()``.
        models: Models namespace with ``list()``.
    """

    def __init__(self, responses: list[MockResponse] | None = None) -> None:
        self.calls: list[MockCall] = []
        self._responses = responses or [MockResponse()]
        self.chat = _MockChatNamespace(self._responses, self.calls)
        self.models = _MockModelsNamespace()

    def __enter__(self) -> MockClient:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> bool:
        return False

    async def __aenter__(self) -> MockClient:
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> bool:
        return False

    def pool_status(self, pool: str | None = None) -> dict:
        """Return mock pool status."""
        return {"mock-pool": {"active": 1, "standby": 0, "total": 1, "current_model": "mock-model"}}

    def active_providers(self) -> list[str]:
        """Return mock provider list."""
        return ["mock-provider"]

    def describe(self, pool: str | None = None) -> str:
        """Return mock description."""
        return 'Pool "mock-pool" (strategy: mock)\n  -> mock-model [mock-provider] (active)'

    def explain(self, *, model: str = "test", **kwargs) -> dict:
        """Return mock routing explanation."""
        return {
            "pool_name": "mock-pool",
            "strategy": "mock",
            "capability": "mock",
            "selected_model": "mock-model",
            "candidates": [{"model_id": "mock-model", "provider_id": "mock-provider", "status": "active"}],
            "reason": "Mock selection",
        }


def mock_client(
    responses: list[MockResponse] | None = None,
) -> MockClient:
    """Create a mock MeshClient for testing.

    Args:
        responses: Pre-configured responses to cycle through.
            Each call to ``chat.completions.create()`` returns
            the next response in the list.

    Returns:
        A ``MockClient`` that can be used in place of a real
        ``MeshClient``.

    Example::

        client = mock_client(responses=[
            MockResponse(content="Hello!"),
            MockResponse(content="Goodbye!"),
        ])
        resp = client.chat.completions.create(model="test", messages=[...])
        assert resp.choices[0].message.content == "Hello!"
    """
    return MockClient(responses=responses)


__all__ = [
    "MockCall",
    "MockClient",
    "MockResponse",
    "mock_client",
]
